client/add/add.ctrl.js

```diff
@@ -19,85 +19,111 @@ module.exports = class Add {
     this.$mdToast = $mdToast;
     this.errorHandler = errorHandler;
   }
-  box (event) {
-    const useFullScreen
-      = (this.$mdMedia('sm') || this.$mdMedia('xs')) && this.$scope.customFullscreen;
+
+  useFullScreen() {
+    return (this.$mdMedia('sm') || this.$mdMedia('xs')) && this.$scope.customFullscreen;
+  }
+
+  watchFullScreen() {
     this.$scope.$watch(() => {
       return this.$mdMedia('xs') || this.$mdMedia('sm');
     }, wantsFullScreen => {
       this.$scope.customFullscreen = (wantsFullScreen === true);
     });
+  }
+
+  dialog(controller, view, locals, event) {
     return Promise.resolve(this.$mdDialog.show({
-      controller: ['$mdDialog', boxCtrl],
+      controller: controller,
       controllerAs: 'dialog',
-      templateUrl: 'add/box.view.html',
+      templateUrl: view,
       parent: angular.element(document.body),
       targetEvent: event,
       clickOutsideToClose: true,
-      fullscreen: useFullScreen,
+      fullscreen: this.useFullScreen(),
       bindToController: true,
-      locals: {
-        defaultBoxVisibility: this.prefs.defaultBoxVisibility
-      }
-    }).then((boxInfo) => this.io.socket.postAsync('/api/v1/box', boxInfo)))
+      locals
+    }));
+  }
+
+  toast(message, action) {
+    const toast = this.$mdToast
+      .simple()
+      .position('bottom right')
+      .hideDelay(4000)
+      .textContent(message)
+      .action(action);
+    return this.$mdToast.show(toast);
+  }
+
+  toastNoHide(message, action) {
+    return this.$mdToast
+      .simple()
+      .position('bottom right')
+      .hideDelay(false)
+      .textContent(message)
+      .action(action);
+  }
+
+  addedToast(message, action, location) {
+    this.toast(message, action)
+      .then(checkOk)
+      .then(() => this.$location.path(location))
+      .catch(this.errorHandler);
+  }
+
+  box (event) {
+    this.watchFullScreen();
+    const locals = {
+      defaultBoxVisibility: this.prefs.defaultBoxVisibility
+    };
+    return this.dialog(['$mdDialog', boxCtrl], 'add/box.view.html', locals, event)
+      .then(boxInfo => {
+        const toast = this.toastNoHide('Creating box');
+        return Promise.any([
+          this.$mdToast.show(toast),
+          this.io.socket.postAsync('/api/v1/box', boxInfo)
+        ]).tap(() => this.$mdToast.hide(toast));
+      })
       .tap(box => {
```

Comment:
It waits for at least 1 of the toast being hidden or the socket call to finish, and then it'll run. We return the Promise.any. But yeah, I don't allow the toast to hide itself.

Having said that, maybe it doesn't matter, I don't actually care about whether the toast hides, because if it can't hide itself, I'm only caring about the socket finishing. So that's unnecessary.